onnxruntime/contrib_ops/cuda/bert/attention_impl.cu

```diff
@@ -334,55 +334,132 @@ Status PrepareQkv(contrib::AttentionParameters& parameters,
+      else if (data.past_key == nullptr && data.present_key != nullptr) {
+        assert(data.past_value == nullptr);
+        assert(data.present_value != nullptr);
+        assert(data.query != nullptr);
+        assert(data.key != nullptr);
+        assert(data.value != nullptr);
+
+        // TODO: supporting packed qkv for self attention may benefit performance
+        ORT_RETURN_IF_ERROR(LaunchTransQkv(stream, 1, sequence_length, batch_size, qk_head_size, num_heads,
+                            max_threads_per_block, false, data.query, q));
+
+        // TODO: supporting packed kv for cross attention may benefit performance
+        ORT_RETURN_IF_ERROR(LaunchTransQkv(stream, 1, kv_sequence_length, batch_size, qk_head_size, num_heads,
+                            max_threads_per_block, false, data.key, data.present_key));
+        ORT_RETURN_IF_ERROR(LaunchTransQkv(stream, 1, kv_sequence_length, batch_size, v_head_size, num_heads,
+                            max_threads_per_block, false, data.value, data.present_value));
+      }
+      // self attention with past and present state
+      else {
+        assert(data.past_key != nullptr);
+        assert(data.past_value != nullptr);
+        assert(data.present_key != nullptr);
+        assert(data.present_value != nullptr);
+        assert(data.query != nullptr);
+        assert(data.key != nullptr);
+        assert(data.value != nullptr);
+        // TODO: supporting packed qkv for self attention may benefit performance
+        ORT_RETURN_IF_ERROR(LaunchTransQkv(stream, 1, sequence_length, batch_size, qk_head_size, num_heads,
+                            max_threads_per_block, false, data.query, q));
+        ORT_RETURN_IF_ERROR(LaunchTransQkv(stream, 1, kv_sequence_length, batch_size, qk_head_size, num_heads,
+                            max_threads_per_block, false, data.key, k));
+        ORT_RETURN_IF_ERROR(LaunchTransQkv(stream, 1, kv_sequence_length, batch_size, v_head_size, num_heads,
+                            max_threads_per_block, false, data.value, v));
+      }
+      qkv_format = AttentionQkvFormat::Q_K_V_BNSH;
     }
-    // cross attention with present state or self attention with present state
-    else if (data.past_key == nullptr && data.present_key != nullptr) {
-      assert(data.past_value == nullptr);
-      assert(data.present_value != nullptr);
-      assert(data.query != nullptr);
-      assert(data.key != nullptr);
-      assert(data.value != nullptr);
-
-      // TODO: supporting packed qkv for self attention may benefit performance
-      ORT_RETURN_IF_ERROR(LaunchTransQkv(stream, 1, sequence_length, batch_size, qk_head_size, num_heads,
-                          max_threads_per_block, false, data.query, q));
-
-      // TODO: supporting packed kv for cross attention may benefit performance
+#if USE_FLASH_ATTENTION
+    else if (use_memory_efficient_attention && data.past_key != nullptr && data.past_value != nullptr && parameters.pass_past_in_kv) {
```

Comment:
We can also consider transpose BNSH to BSNH and pass them through data.key and data.value. Which one is better?

Reply:
Discussed offline